### DISCUSSION TITLE
Exithandler with email/slack push

### DISCUSSION THREAD

**Discussion by ianbenlolo** [kubeflow/pipelines#10955] [General]
Hey,

This may be a duplicate but I have not found this exact question so apologies in advance.

I have a pipeline that i would like to add an exithandler that runs some cleanup code and sends an email.
This is how I've done it so far:

```
@dsl.component
def dummy_print_action(action: str):
    print(action)

@dsl.pipeline
def send_failure_email(
    email_recipients
):
    # this was a hack required to use the VertexNotificationEmailOp component only when the pipeline fails
    exit_handler_task = vertex_notification_email.VertexNotificationEmailOp(recipients=email_recipients)

    with dsl.ExitHandler(exit_task=exit_handler_task, name="Dummy exit handler"):
        _ = dummy_print_action(
                action=f"Failed pipeline, sending notification email to {email_recipients}"
            )
@dsl.pipeline
def exit_handler(
    email_recipients: List[str],
    status: dsl.PipelineTaskFinalStatus,
):
    get_run_state_task = get_run_state(status=status)

    with dsl.Condition(
        get_run_state_task.output != "SUCCEEDED", name="state != SUCCEEDED"
    ):
        _ = cleanup()
        _ = send_failure_email(
            email_recipients=email_recipients,
        )

@dsl.pipeline
def pipeline(args):
    exit_handler_task = exit_handler(email_recipients=["my@email.com"],)
    with dsl.ExitHandler(exit_task=exit_handler_task, name="Pipeline"):
        pipeline_stuff()
```

But this is a little cumbersome. I was wondering if there is an easier way to do [this](https://stackoverflow.com/questions/57508382/kubeflow-pipeline-termination-notificaiton)?

Furthermore, i'd like to send a slack notification on failure and haven't found a way to other than the link but I'm unsure how to set up the necessary image.

Thank you!

**Comment by chensun** [comment]
I think you don't need the dummy exit handler part. Can't you use `VertexNotificationEmailOp` directly right after the cleanup()?

**Reply by ianbenlolo** [reply]
Hey! Thanks your your reply. 

No because it can only be used in an exit handler (see [docs](https://google-cloud-pipeline-components.readthedocs.io/en/google-cloud-pipeline-components-1.0.26/google_cloud_pipeline_components.v1.vertex_notification_email.html#google_cloud_pipeline_components.v1.vertex_notification_email.VertexNotificationEmailOp)).
